src/Containers/Containers.jl

```diff
@@ -0,0 +1,226 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#############################################################################
+# JuMP
+# An algebraic modeling language for Julia
+# See http://github.com/JuliaOpt/JuMP.jl
+#############################################################################
+
+"""
+    Containers
+
+Module defining the containers `SparseAxisArray` that behaves as a regular
+`AbstractArray` but with custom indexes that are not necessarily integers.
+"""
+module Containers
+
+"""
+    struct SparseAxisArray{T,N,K<:NTuple{N, Any}} <: AbstractArray{T,N}
+        data::Dict{K,T}
+    end
+
+`N`-dimensional array with elements of type `T` where only a subset of the
+entries are defined. The entries with indices `idx = (i1, i2, ..., iN)` in
+`keys(data)` has value `data[idx]`. Note that as opposed to
+`SparseArrays.AbstractSparseArray`, the missing entries are not assumed to be
+`zero(T)`, they are simply not part of the array. This means that the result of
+`map(f, sa::SparseAxisArray)` or `f.(sa::SparseAxisArray)` has the same sparsity
+structure than `sa` even if `f(zero(T))` is not zero.
+"""
+struct SparseAxisArray{T,N,K<:NTuple{N, Any}} <: AbstractArray{T,N}
+    data::Dict{K,T}
+end
+
+Base.length(sa::SparseAxisArray) = length(sa.data)
+Base.IteratorSize(::Type{<:SparseAxisArray}) = Base.HasLength()
+# By default `IteratorSize` for `Generator{<:AbstractArray{T,N}}` is
+# `HasShape{N}`
+Base.IteratorSize(::Type{Base.Generator{<:SparseAxisArray}}) = Base.HasLength()
+# Needed in `collect_to_with_first!`
+Base.eachindex(g::Base.Generator{<:SparseAxisArray}) = eachindex(g.iter)
+Base.iterate(sa::SparseAxisArray, args...) = iterate(values(sa.data), args...)
+# Used by JuMP.getloopedcode
```

Comment:
I don't think the comments on eachindex and haskey are needed. These methods will be used in multiple places, and the comment will go out of date.